### ISSUE TITLE
Add support for NoSQL Bench for Cassandra cluster testing 

### ISSUE THREAD

**Issue by pmcfadin** [k8ssandra/k8ssandra#1054]
I'm using this issue to persist a discussion that happened in Discord. There was a discussion around support for NoSQL Bench (https://github.com/nosqlbench/nosqlbench) in k8ssandra to do testing like described here: https://k8ssandra.io/blog/articles/k8ssandra-performance-benchmarks-on-cloud-managed-kubernetes/ 

I can certainly see the benefit of having a benchmarking tool available for operators as they deploy K8ssandra clusters on various hardware. The split I can see on a solution like in a couple of options. 

1. Inside k8ssandra as an option but not default
2. A separate repo inside of the k8ssandra org

In either case, the ability to create a testing namespace for NoSQL Bench that has access to the running Cassandra cluster in the same Kube cluster.

**Comment by Miles-Garnsey** [comment]
So I think that Nosqlbench is already 'supported' in some sense - insofar as if you deploy Nosqlbench alongside k8ssandra/cass-operator, there is nothing that will stop it working.

With that in view, I'd propose a third option: provide documentation showing how to deploy Nosqlbench within a standard k8s deployment, and how to operate it by uploading the workload descriptions as `configMaps` or via `kubectl cp`.

We have to be careful not to add every possible application into k8ssandra to keep things lean and comprehensible. But there is nothing to stop us from fleshing out documentation with details on how to do integration between the k8ssandra components and other common things in the ecosystem.

**Comment by jdonenine** [comment]
Would be interesting to get some ideas here from @grighetto since he's done the work of deploying NoSQLBench in k8s for testing within Adelphi.

**Comment by Miles-Garnsey** [comment]
Just off the top of my head (because I've done a fair bit of work in this space recently) to make this fly we would need - 

* A deployment for Nosqlbench (simple)
* Some ConfigMap or `kubectl cp` stuff to get the workload into the deployment.
* A bit of work around getting the Grafana dashboards into Grafana. It looks like we're using a particular [sidecar](https://github.com/kiwigrid/k8s-sidecar) to scrape all ConfigMaps with a specified label and place them in the Grafana container. I don't mind this solution as the alternative would be the Grafana operator (which seems like overkill).
* A bit of work to ensure that Prometheus knows about the Nosqlbench instance. I think this is just another `servicemonitors.monitoring.coreos.com` CR. 

So all told, there's maybe three manifests required to get the integration sorted. But I'm keen to hear from @grighetto as I may have missed something here.

**Comment by grighetto** [comment]
@Miles-Garnsey Right, that's pretty similar to what we have in [Adelphi](https://github.com/datastax/adelphi) right now.
I'd just add that NoSQLBench doesn't talk to Prometheus directly. NB pushes the metrics to an intermediate service in Graphite format, which then converts them to Prom format and exposes them as web service.
That intermediate service is this graphite-exporter: https://hub.docker.com/r/prom/graphite-exporter

NoSQLBench has a sample setup using docker containers locally, it does so programmatically like this: https://github.com/nosqlbench/nosqlbench/blob/main/engine-docker/src/main/java/io/nosqlbench/engine/docker/DockerMetricsManager.java 

And I "converted" that into Kubernetes templates:

**Graphite exporter**
https://github.com/datastax/adelphi/blob/master/helm/adelphi/templates/graphite.yaml
**Graphite exporter svc**
https://github.com/datastax/adelphi/blob/master/helm/adelphi/templates/graphite-service.yaml
**Graphite exporter configmap** (for mapping the metrics to the right format)
https://github.com/datastax/adelphi/blob/master/helm/adelphi/templates/graphite-configmap.yaml

**Prometheus configmap** (to point it to the graphite service)
https://github.com/datastax/adelphi/blob/master/helm/adelphi/templates/prometheus-configmap.yaml

**Grafana dashboard config** 
https://github.com/datastax/adelphi/blob/master/helm/adelphi/config/grafana/nb4-new-dashboard.json

And then when you run a NoSQLBench job, you just have to point it to the graphite service like this:
https://github.com/datastax/adelphi/blob/master/helm/adelphi/templates/nosqlbench-jobs.yaml#L16

**Comment by bradfordcp** [comment]
I'm curious if some of the work recently done by @omnifroodle could be pulled in to satisfy this ticket. Is this something we want to bake into the project or document as a task on the website?

**Comment by jsanda** [comment]
I'd say no to baked into the project, yes to some docs
